handlers/jiochat/jiochat.go

```diff
@@ -0,0 +1,323 @@
+
+		err := h.Backend().WriteChannelEvent(ctx, channelEvent)
+		if err != nil {
+			return nil, err
+		}
+
+		return []courier.Event{channelEvent}, courier.WriteChannelEventSuccess(ctx, w, r, channelEvent)
+	}
+
+	if jcRequest.MsgType == "event" {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("unknown event"))
+	}
+
+	// create our message
+	msg := h.Backend().NewIncomingMsg(channel, urn, jcRequest.Content).WithExternalID(fmt.Sprintf("%d", jcRequest.MsgID)).WithReceivedOn(date)
+	if jcRequest.MsgType == "image" || jcRequest.MsgType == "video" || jcRequest.MsgType == "voice" {
+		mediaURL := resolveMediaID(jcRequest.MediaID)
+		msg.WithAttachment(mediaURL)
+	}
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+}
+
+func resolveMediaID(mediaID string) string {
+	mediaURL, _ := url.Parse(fmt.Sprintf("%s/%s", jiochatURL, "media/download.action"))
+	mediaURL.RawQuery = url.Values{"media_id": []string{mediaID}}.Encode()
+	return mediaURL.String()
+}
+
+type jiochatAccess interface {
+	fetchAccessToken(*handler, courier.Channel) error
+	getAccessToken(*handler, courier.Channel) string
+}
+
+type jiochatClient struct{}
+
+type refreshTokenData struct {
+	GrantType    string `json:"grant_type"`
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+func (jCli *jiochatClient) fetchAccessToken(h *handler, channel courier.Channel) error {
+	time.Sleep(jiochatFetchTimeout)
+
+	tokenURL, _ := url.Parse(fmt.Sprintf("%s/%s", jiochatURL, "auth/token.action"))
+
+	refreshTokenData := &refreshTokenData{
+		GrantType:    "client_credentials",
+		ClientID:     channel.StringConfigForKey(configJiochatAppID, ""),
+		ClientSecret: channel.StringConfigForKey(configJiochatAppSecret, ""),
+	}
+
+	jsonBody, err := json.Marshal(refreshTokenData)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, tokenURL.String(), bytes.NewReader(jsonBody))
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	rr, err := utils.MakeHTTPRequest(req)
+
+	accessToken, err := jsonparser.GetString([]byte(rr.Body), "access_token")
+	if err != nil {
+		return err
+	}
+
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+	cacheKey := fmt.Sprintf("jiochat_channel_access_token:%s", channel.UUID().String())
+
+	_, err = rc.Do("set", cacheKey, accessToken, 7200)
+	return err
+}
+
+func (jCli *jiochatClient) getAccessToken(h *handler, channel courier.Channel) string {
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+	cacheKey := fmt.Sprintf("jiochat_channel_access_token:%s", channel.UUID().String())
+	accessToken, _ := redis.String(rc.Do("GET", cacheKey))
+	return accessToken
+}
+
+func (h *handler) findAccessToken(channel courier.Channel) string {
+	return h.jiochatClient.getAccessToken(h, channel)
```

Comment:
Single line methods like this are anti-patterns, why even have it?

Reply:
this is influenced by the interface to be able to test